### PR TITLE
Improve const correctness of hypre interface [hypre-const]

### PR DESCRIPTION
**Pull request by pazner**
In the spirit of improving MFEM's const-correctness, this PR adds const qualifiers to several member functions and parameters in the hypre interface. The changes should all be non-breaking, with the exception of the `GetData` member functions of `HypreDiagScale` and `HypreTriSolve`, which returned non-const pointers to the underlying matrix. I changed those to use `const_cast` and marked them as deprecated — I don't think they are widely used (if at all). const-correct versions of those two functions are added as well.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2197](https://github.com/mfem/mfem/pull/2197) | @pazner | @tzanio | @YohannDudouit + @jandrej + @mlstowell | 04/24/21 | 04/30/21 | 05/04/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by jandrej — APPROVED** [review]

**Comment by jandrej** [comment]
LGTM

File: linalg/hypre.hpp [hunk 1]
```diff
@@ -838,15 +838,20 @@ class HypreTriSolve : public HypreSolver
...
 
    virtual HYPRE_PtrToParSolverFcn SetupFcn() const
    { return (HYPRE_PtrToParSolverFcn) HYPRE_ParCSROnProcTriSetup; }
    virtual HYPRE_PtrToParSolverFcn SolveFcn() const
    { return (HYPRE_PtrToParSolverFcn) HYPRE_ParCSROnProcTriSolve; }
 
-   HypreParMatrix* GetData() { return A; }
+   const HypreParMatrix* GetData() const { return A; }
+
+   /// Deprecated. Use %ref HypreTriSolve::GetData() const
```
Comment (v-dobrev) [inline linalg/hypre.hpp:851]: The `%ref` here looks like a typo -- you probably meant `@ref` or `\ref`. Note that the link to the following method will be created even without `@ref` or `\ref`.

There is at least one more place below where I saw `%ref` in the doxygen comments.

Comment (pazner) [reply linalg/hypre.hpp]: Good catch. Fixed.

**Comment by mlstowell** [comment]
I noticed a few other methods we might make `const`:
`HypreParVector::GlobalSize()`
`HypreParMatrix::Print()`

We might even consider the two `HypreParMatrix::Mult` methods which are not yet `const`. These are called by the other Mult methods which are already `const` so this would be consistent.

Also, we could make `HypreLOBPCG::GetEigenvalues()` and `HypreLOBPCG::GetEigenVector()` const. The similar methods in `HypreAME` could also be `const` but this would require a little more effort since they copy precomputed data from HYPRE structs to member variables local to the class.

It might make sense to save the eigenvalue solver changes for another PR but the other changes might fit here.

**Review by mlstowell — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: linalg/hypre.hpp [hunk 2]
```diff
@@ -119,18 +119,18 @@ class HypreParVector : public Vector
...
+   MPI_Comm GetComm() const { return x->comm; }
 
    /// Converts hypre's format to HypreParVector
    void WrapHypreParVector(hypre_ParVector *y, bool owner=true);
 
    /// Returns the parallel row/column partitioning
    /** See @ref hypre_partitioning_descr "here" for a description of the
        partitioning array. */
-   inline HYPRE_BigInt *Partitioning() { return x->partitioning; }
+   inline HYPRE_BigInt *Partitioning() const { return x->partitioning; }
```
Comment (YohannDudouit) [inline linalg/hypre.hpp:130]: If we want this method to be `const` the returned internal should be `const` too I think:
```suggestion
   inline const HYPRE_BigInt *Partitioning() const { return x->partitioning; }
```

**Review by YohannDudouit — APPROVED** [review]

File: linalg/hypre.hpp (hunk 2 above)
Comment (pazner) [reply linalg/hypre.hpp:130]: I agree. Good catch.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: linalg/hypre.hpp (hunk 2 above)
Comment (v-dobrev) [reply linalg/hypre.hpp]: Adding a const version is fine but why remove the non-const?

Comment (pazner) [reply linalg/hypre.hpp]: Is modifying the partitioning array even something that is a well-defined operation after construction of the vector?

Comment (v-dobrev) [reply linalg/hypre.hpp]: It's more of a convenience thing -- you may want to get the partition to pass it to some hypre function, avoiding the `const_cast` if your hypre vector is not `const`.

Comment (pazner) [reply linalg/hypre.hpp]: If that's something useful that users really do in their code then we can definitely bring back the non-const accessor (even without it you can get the non-const partitioning by casting to `hypre_ParVector*` and accessing the `partitioning` field directly if needed).

Do you know if this is something that exists in the wild somewhere?

Comment (v-dobrev) [reply linalg/hypre.hpp]: It is one line and preserves backward compatibility -- I see no reason to remove it.

Comment (pazner) [reply linalg/hypre.hpp]: The reason would be to disallow the user from mutating the partitioning of an already created vector (which in my understanding is not a valid thing to do — but I may be wrong about this).

Comment (v-dobrev) [reply linalg/hypre.hpp]: For me, this is not a good enough reason to break backward compatibility.

Comment (pazner) [reply linalg/hypre.hpp]: Ok, that's fair, and we can add the non-const version (perhaps as `MFEM_DEPRECATED`?) but I still think it's relevant to know if there is any code out there that depends on this, or whether it's purely theoretical.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
LGTM

File: linalg/hypre.hpp (hunk 2 above)
Comment (YohannDudouit) [reply linalg/hypre.hpp]: "Convenience thing" is the enemy of good software... We don't want to write a software that's easy to hack in, we want something hard to use incorrectly... I think the `const_cast` is actually good, it pushes a user to think about what he's actually doing. It should feel like a hack to modify directly the partitioning array outside its class. If the partitioning array needs to be modified, then there is most likely a better abstraction for that than giving away the internals.

**Pull request merged**